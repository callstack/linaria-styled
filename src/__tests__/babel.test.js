/* @flow */

const babel = require('@babel/core');
const dedent = require('dedent');

const transpile = async input => {
  const { code } = await babel.transformAsync(input, {
    presets: [require.resolve('../babel')],
    filename: '/app/index.js',
  });

  return code;
};

it('transpiles styled template literal', async () => {
  const code = await transpile(
    dedent`
    const Title = styled('h1')\`
      font-size: 14px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('evaluates and inlines expressions in scope', async () => {
  const code = await transpile(
    dedent`
    const color = 'blue';

    const Title = styled('h1')\`
      color: ${'${color}'};
      width: ${'${100 / 3}'}%;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('replaces unknown expressions with CSS custom properties', async () => {
  const code = await transpile(
    dedent`
    const Title = styled('h1')\`
      font-size: ${'${size}'}px;
      color: ${'${props => props.color}'}
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('uses the same custom property for the same identifier', async () => {
  const code = await transpile(
    dedent`
    const Box = styled('div')\`
      height: ${'${size}'}px;
      width: ${'${size}'}px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('uses the same custom property for the same expression', async () => {
  const code = await transpile(
    dedent`
    const Box = styled('div')\`
      height: ${'${props => props.size}'}px;
      width: ${'${props => props.size}'}px;
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('handles nested blocks', async () => {
  const code = await transpile(
    dedent`
    const Button = styled('button')\`
      font-family: ${'${regular}'};

      &:hover {
        border-color: blue;
      }

      @media (max-width: 200px) {
        width: 100%;
      }
    \`;
    `
  );

  expect(code).toMatchSnapshot();
});

it('prevents class name collision', async () => {
  const code = await transpile(
    dedent`
    const Title = styled('h1')\`
      font-size: ${'${size}'}px;
      color: ${'${props => props.color}'}
    \`;

    function more() {
      const Title = styled('h1')\`
        font-family: ${'${regular}'};
      \`;
    }
    `
  );

  expect(code).toMatchSnapshot();
});

it('handles component in object property', async () => {
  const code = await transpile(
    dedent`
    const components = {
      title: styled('h1')\`
        font-size: 14px;
      \`
    };
    `
  );

  expect(code).toMatchSnapshot();
});

it('throws when not attached to a variable', async () => {
  expect.assertions(1);

  try {
    await transpile(
      dedent`
      styled('h1')\`
        font-size: ${'${size}'}px;
        color: ${'${props => props.color}'}
      \`;
      `
    );
  } catch (e) {
    expect(e.message).toMatchSnapshot();
  }
});

it('does not output CSS if none present', async () => {
  const code = await transpile(
    dedent`
      const number = 42;

      const title = String.raw\`This is something\`;
      `
  );

  expect(code).toMatchSnapshot();
});
