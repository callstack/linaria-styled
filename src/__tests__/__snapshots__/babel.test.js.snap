// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not output CSS if none present 1`] = `
"const number = 42;
const title = String.raw\`This is something\`;"
`;

exports[`evaluates and inlines expressions in scope 1`] = `
"const color = 'blue';
const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_t1ugh8t9\\"
});

/*
CSS OUTPUT TEXT START
.Title_t1ugh8t9{color:blue;width:33.333333333333336%;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":3,\\"column\\":6},\\"name\\":\\"Title_t1ugh8t9\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`handles component in object property 1`] = `
"const components = {
  title: styled.component('h1', {
    name: \\"title\\",
    class: \\"title_t1ugh8t9\\"
  })
};

/*
CSS OUTPUT TEXT START
.title_t1ugh8t9{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":2,\\"column\\":2},\\"name\\":\\"title_t1ugh8t9\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`handles nested blocks 1`] = `
"const Button = styled.component('button', {
  name: \\"Button\\",
  class: \\"Button_b1ugh8t9\\",
  vars: {
    \\"b1ugh8t9-0-0\\": regular
  }
});

/*
CSS OUTPUT TEXT START
.Button_b1ugh8t9{font-family:var(--b1ugh8t9-0-0);}.Button_b1ugh8t9:hover{border-color:blue;}@media (max-width:200px){.Button_b1ugh8t9{width:100%;}}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Button_b1ugh8t9\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`prevents class name collision 1`] = `
"const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_t1ugh8t9\\",
  vars: {
    \\"t1ugh8t9-0-0\\": size,
    \\"t1ugh8t9-0-1\\": props => props.color
  }
});

function more() {
  const Title = styled.component('h1', {
    name: \\"Title\\",
    class: \\"Title_t1ugh8t9x\\",
    vars: {
      \\"t1ugh8t9x-1-0\\": regular
    }
  });
}

/*
CSS OUTPUT TEXT START
.Title_t1ugh8t9{font-size:var(--t1ugh8t9-0-0)px;color:var(--t1ugh8t9-0-1);}
.Title_t1ugh8t9x{font-family:var(--t1ugh8t9x-1-0);}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_t1ugh8t9\\"},{\\"generated\\":{\\"line\\":2,\\"column\\":0},\\"original\\":{\\"line\\":7,\\"column\\":8},\\"name\\":\\"Title_t1ugh8t9x\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`replaces unknown expressions with CSS custom properties 1`] = `
"const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_t1ugh8t9\\",
  vars: {
    \\"t1ugh8t9-0-0\\": size,
    \\"t1ugh8t9-0-1\\": props => props.color
  }
});

/*
CSS OUTPUT TEXT START
.Title_t1ugh8t9{font-size:var(--t1ugh8t9-0-0)px;color:var(--t1ugh8t9-0-1);}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_t1ugh8t9\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`throws when not attached to a variable 1`] = `
"/app/index.js: Couldn't determine a name for the component. Ensure that it's either:
- Assigned to a variable
- Is an object property

[0m[31m[1m>[22m[39m[90m 1 | [39mstyled([32m'h1'[39m)[32m\`[39m[0m
[0m [90m   | [39m[31m[1m^[22m[39m[0m
[0m [90m 2 | [39m[32m  font-size: \${size}px;[39m[0m
[0m [90m 3 | [39m[32m  color: \${props => props.color}[39m[0m
[0m [90m 4 | [39m[32m\`[39m[33m;[39m[0m"
`;

exports[`transpiles styled template literal 1`] = `
"const Title = styled.component('h1', {
  name: \\"Title\\",
  class: \\"Title_t1ugh8t9\\"
});

/*
CSS OUTPUT TEXT START
.Title_t1ugh8t9{font-size:14px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Title_t1ugh8t9\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`uses the same custom property for the same expression 1`] = `
"const Box = styled.component('div', {
  name: \\"Box\\",
  class: \\"Box_b1ugh8t9\\",
  vars: {
    \\"b1ugh8t9-0-0\\": props => props.size
  }
});

/*
CSS OUTPUT TEXT START
.Box_b1ugh8t9{height:var(--b1ugh8t9-0-0)px;width:var(--b1ugh8t9-0-0)px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Box_b1ugh8t9\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;

exports[`uses the same custom property for the same identifier 1`] = `
"const Box = styled.component('div', {
  name: \\"Box\\",
  class: \\"Box_b1ugh8t9\\",
  vars: {
    \\"b1ugh8t9-0-0\\": size
  }
});

/*
CSS OUTPUT TEXT START
.Box_b1ugh8t9{height:var(--b1ugh8t9-0-0)px;width:var(--b1ugh8t9-0-0)px;}

CSS OUTPUT TEXT END

CSS OUTPUT MAPPINGS:[{\\"generated\\":{\\"line\\":1,\\"column\\":0},\\"original\\":{\\"line\\":1,\\"column\\":6},\\"name\\":\\"Box_b1ugh8t9\\"}]
CSS OUTPUT DEPENDENCIES:[]
*/"
`;
